src/CodeGen_WebGPU_Dev.cpp

```diff
@@ -0,0 +1,385 @@
+               << args_var << " : " << struct_name << " ;\n\n";
+    }
+
+    // Determine the workgroup size.
+    // TODO: Remove this and use overridable constants instead.
+    struct GetWorkgroupSize : public IRVisitor {
+        using IRVisitor::visit;
+        void visit(const For *loop) override {
+            if (!is_gpu_var(loop->name)) {
+                return loop->body.accept(this);
+            }
+            if (loop->for_type != ForType::GPUThread) {
+                return loop->body.accept(this);
+            }
+            int index = loop_name_to_wgsize_index(loop->name);
+            user_assert(index >= 0 && index <= 2)
+                << "invalid 'wgsize' index for loop variable '" << loop->name
+                << "'.\n";
+            const IntImm *limit = loop->extent.as<IntImm>();
+            user_assert(limit != nullptr)
+                << "dynamic workgroup sizes are not yet supported.";
+            user_assert(limit->value > 0)
+                << "'" << loop->name << "' must be greater than zero.\n";
+            values[index] = limit->value;
+            debug(4) << "wgsize[" << index << "] is " << values[index] << "\n";
+            loop->body.accept(this);
+        }
+        int loop_name_to_wgsize_index(const string &name) {
+            string ids[] = {
+                ".__thread_id_x",
+                ".__thread_id_y",
+                ".__thread_id_z",
+                ".__thread_id_w",
+            };
+            for (auto &id : ids) {
+                if (ends_with(name, id)) {
+                    return (&id - ids);
+                }
+            }
+            return -1;
+        }
+        int values[3] = {1, 1, 1};
+    };
+    GetWorkgroupSize wgsize;
+    s.accept(&wgsize);
+
+    // Emit the function prototype.
+    stream << "[[stage(compute), workgroup_size("
+           << wgsize.values[0] << ", "
+           << wgsize.values[1] << ", "
+           << wgsize.values[2] << ")]]\n";
+    stream << "fn " << name << "(\n"
+           << "  [[builtin(local_invocation_id)]] local_id : vec3<u32>,\n"
+           << "  [[builtin(workgroup_id)]] group_id : vec3<u32>,\n"
+           << ")\n";
+
+    open_scope();
+
+    // Redeclare non-buffer arguments at function scope.
+    for (size_t i = 0; i < args.size(); i++) {
+        if (!args[i].is_buffer) {
+            stream << get_indent() << "let " << print_name(args[i].name)
+                   << " = " << args_var << "."
+                   << print_name(args[i].name) << ";\n";
+        }
+    }
+
+    // Generate function body.
+    print(s);
+
+    close_scope("shader " + name);
+}
+
+void CodeGen_WebGPU_Dev::CodeGen_WGSL::visit(const Cast *op) {
+    print_assignment(op->type,
+                     print_type(op->type) + "(" + print_expr(op->value) + ")");
+}
+
+void CodeGen_WebGPU_Dev::CodeGen_WGSL::visit(const IntImm *op) {
+    internal_assert(op->type.bits() == 32)
+        << "WGSL only supports 32-bit integers";
```

Comment:
...well that is disappointing. No other int sizes? or unsigned?

Reply:
All I can say is that most of the high-performance, well-scheduled chunks of Halide code I've seen in production make *extensive* use of 8- and 16-bit vector types. Their absence would be a big limitation for maximizing performance (and memory usage), not to mention making portable pipelines a lot more painful to write and schedule (since pipelines running on typical SIMD or DSP architectures won't want to leave the benefits of having smaller ints, and thus effectively-wider vectors, on the table...)

Reply:
For other type-challenged backends (d3d12) I believe we emulate low bit width types in Halide, so you could take a look at those backends. Feel free to email me for help demonstrating enough demand.